### PR TITLE
Value cache

### PR DESCRIPTION
**Pull request by aberaud**

### CODE REVIEW

File: src/dht_proxy_client.cpp [hunk 1]
```diff
@@ -426,61 +450,82 @@ DhtProxyClient::getPublicAddress(sa_family_t family)
...
 
 size_t
-DhtProxyClient::listen(const InfoHash& key, GetCallback cb, Value::Filter filter, Where where)
+DhtProxyClient::listen(const InfoHash& key, GetCallback cb, Value::Filter filter, Where where) {
+    auto it = listeners_.find(key);
+    if (it == listeners_.end()) {
+        it = listeners_.emplace(key, ProxySearch{/*
+            ValueCache([this,key](const std::vector<Sp<Value>>& values, bool expired){
+                std::cout << "Listen: callback with " << values.size() << " values" << std::endl;
+                auto s = listeners_.find(key);
```
Comment (AmarOk1412) [inline src/dht_proxy_client.cpp]: to remove

File: src/dht_proxy_client.cpp [hunk 2]
```diff
@@ -426,61 +450,82 @@ DhtProxyClient::getPublicAddress(sa_family_t family)
...
-    l.filterChain = std::move(filterChain);
-    l.isCanceledViaClose = isCanceled;
-    std::weak_ptr<bool> isCanceledViaClose(l.isCanceledViaClose);
-    l.thread = std::thread([=]()
+    auto pushNotifToken = std::make_shared<unsigned>(0);
+    l->second.pushNotifToken = pushNotifToken;
+    //l.callbackId = callbackId;
+    //l.pushNotifToken = pushNotifToken;
+    //l.filter = std::move(filter);
+    //l.isCanceledViaClose = std::make_shared<bool>(false);
```
Comment (AmarOk1412) [inline src/dht_proxy_client.cpp]: to remove

**Comment by AmarOk1412** [comment]
I just tried with the proxy without (push notifications) and it worked perfectly. Just need to validate with push.

File: tools/dhtnode.cpp [hunk 3]
```diff
@@ -254,11 +247,7 @@ void cmd_loop(std::shared_ptr<DhtRunner>& dht, dht_params& params
         }
         dht::InfoHash id;
 
-        if (op == "cl") {
-            std::string hash, rem;
-            iss >> hash >> rem;
-            dht->cancelListen(dht::InfoHash(hash), std::stoul(rem));
-        }
+        if (false) {}
```
[the remark this answers is not in the document]
Comment (aberaud) [reply tools/dhtnode.cpp]: https://github.com/savoirfairelinux/opendht/issues/259

File: src/op_cache.h [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        if (op == ops.end()) {
+            // find satisfying query
+            for (auto it = ops.begin(); it != ops.end(); it++) {
+                if (q->isSatisfiedBy(*it->first)) {
+                    op = it;
+                    break;
+                }
+            }
+        }
+        if (op == ops.end()) {
```
[the remark this answers is not in the document]
Comment (aberaud) [reply src/op_cache.h]: https://github.com/savoirfairelinux/opendht/issues/259

File: src/search.h [hunk 5]
```diff
@@ -201,6 +213,17 @@ struct Dht::SearchNode {
         getStatus.clear();
     }
 
+    void onValues(const Sp<Query>& q, net::RequestAnswer&& answer, const TypeStore& types, const time_point& now)
+    {
+        std::cout << "onValues" << std::endl;
```
[the remark this answers is not in the document]
Comment (aberaud) [reply src/search.h]: done

File: include/opendht/dht_proxy_client.h [hunk 6]
```diff
@@ -102,7 +104,7 @@ class OPENDHT_PUBLIC DhtProxyClient final : public DhtInterface {
         get(key, bindGetCb(cb), bindDoneCb(donecb), std::forward<Value::Filter>(f), std::forward<Where>(w));
     }
 
-    void get(const InfoHash& key, const GetCallback& cb, DoneCallback donecb, const Value::Filter& filterChain);
+    //void get(const InfoHash& key, const GetCallback& cb, DoneCallback donecb, const Value::Filter& filterChain);
```
[the remark this answers is not in the document]
Comment (aberaud) [reply include/opendht/dht_proxy_client.h]: done

File: tools/dhtnode.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (aberaud) [reply tools/dhtnode.cpp]: woudn't compile if OPENDHT_INDEXATION is disabled because the first statement would be 'else if'

**Pull request merged**